src/core/error_helpers.js

```diff
@@ -268,68 +296,138 @@ if (typeof IS_MINIFIED !== 'undefined') {
+          break;
         case 'function':
-          return param instanceof Function;
+          matches = param instanceof Function;
+          break;
+        case 'null':
+          matches = param === null;
+          break;
       }
+    } else {
+      matches = typeof param === type.t;
     }
-
-    return typeof param === type.t;
+    return matches ? 0 : 1;
   };
 
   // testType() for non-object type parameter validation
-  // Returns true if PASS, false if FAIL
   var testParamTypes = function(param, types) {
-    for (var i = 0; i < types.length; i++) {
-      if (testParamType(param, types[i])) {
-        return true;
+    var minScore = 9999;
+    for (var i = 0; minScore > 0 && i < types.length; i++) {
+      var score = testParamType(param, types[i]);
+      if (minScore > score) minScore = score;
+    }
+    return minScore;
+  };
+
+  // generate a score (higher is worse) for applying these args to
+  // this overload.
+  var scoreOverload = function(args, argCount, overload, minScore) {
+    var score = 0;
+    var formats = overload.formats;
+
+    // check for too few/many args
+    // the score is double number of extra/missing args
+    if (argCount < overload.minParams) {
+      score = (overload.minParams - argCount) * 2;
+    } else if (argCount > formats.length) {
+      score = (argCount - formats.length) * 2;
+    }
+
+    // loop through the formats, adding up the error score for each arg.
+    // quit early if the score gets higher than the previous best overload.
+    for (var p = 0; score <= minScore && p < formats.length; p++) {
+      var arg = args[p];
+      var format = formats[p];
+      if (arg == null) {
+        // handle non-optional and non-trailing undefined args
+        if (
+          !format.optional ||
+          (p < overload.minParams ? true : p < argCount)
+        ) {
+          score += 1;
+        }
+      } else {
+        score += testParamTypes(arg, format.types);
       }
     }
-    return false;
+    return score;
   };
 
-  var testParamFormat = function(args, formats) {
+  // gets a list of errors for this overload
+  var getOverloadErrors = function(args, argCount, overload) {
+    var formats = overload.formats;
+
+    // check for too few/many args
+    if (argCount < overload.minParams) {
+      return [
+        {
+          type: 'TOO_FEW_ARGUMENTS',
+          argCount: argCount,
+          minParams: overload.minParams
+        }
+      ];
+    } else if (argCount > formats.length) {
+      return [
+        {
+          type: 'TOO_MANY_ARGUMENTS',
+          argCount: argCount,
+          maxParams: formats.length
+        }
+      ];
+    }
+
     var errorArray = [];
     for (var p = 0; p < formats.length; p++) {
       var arg = args[p];
       var format = formats[p];
-      var argType = typeof arg;
-      if ('undefined' === argType || null === arg) {
-        if (format.optional !== true) {
+      if (arg == null) {
```

Comment:
is this supposed to be three equal signs or two? i think do a == test like this can be easily misread or misinterpreted , we could just check `if(!arg)` instead to be more claer

Reply:
yes, the `== null` is intentional and returns a different result than just the falsey check (which would be wrong here).

(sorry if this is duplicate, i wrote this before, but it looks like i deleted it by mistake)

Reply:
can you give me an example of a case where it returns a different result?

Reply:
0, false, ''